Debugger code tracing

TryTwo opened the pull request:
Adds a button to CodeWidget to open a Trace window. Records a section of code between the PC and a breakpoint as it executes, then allows you to follow any register or memory address through it, forwards or backwards. Can click the output to jump to that location in CodeWidget.

For example, if you know r10 has a value for your character's health and you want to know where it is modified or compared at. Normally you may have to dive through multiple branch links and any number of sub branch links to find it, plus register swaps and maybe memory writes and reads. Instead, this will find all references to it for you based on what was executed, while tracking it through register and memory swaps, Does not show actual values of what is being tracked.

Updates GekkoDisassembler with important consistency changes, that allow regex to work its magic.

Updates PPCDebugInterface with code to allow target memory to be calculated from load/store instructions with offsets. I have another, old PR that uses this to let you directly navigate to memory from CodeWidget load/store instructions that use offsets, so you never need to manually calculate an address, but it was never merged.

I'll need someone to confirm it builds okay, because msvc 14.26 bugs out for me.

JosJuice commented:
@dolphin-emu-bot rebuild

JosJuice commented:
When you use the term PC in a translatable string, please add an i18n comment, like here: https://github.com/dolphin-emu/dolphin/blob/62f331dd13e56ad0aea9783fb5fbe5060aa451f3/Source/Core/DolphinQt/ToolBar.cpp#L109

sepalani commented:
@dolphin-emu-bot rebuild

Tilka commented:
Needs a bit of work to fix some stability bug(s?), currently the UI hangs if I click the Run Trace button a couple of times, but I like the idea.

TryTwo commented:
Thanks. I'll disable the buttons while the trace is running, because it can take a few seconds.  I'll see if there are any other issues as well.  I also know I need to fix up the button labels and explanations a bit.

sepalani commented:
I did some testing and it seems the tracing state/button isn't updated properly when switching games.

I also think some of the tracing code should have its own class/file(s) rather than implementing all of it in the GUI code. It can be put in a location such as `Core/Common/Debug` or somewhere more fitting. IMHO, it will provide at least the following benefits:
 - The GUI part will be easier to maintain.
 - It will be easier to reuse the tracing logic.

TryTwo commented on Source/Core/Core/Debugger/PPCDebugInterface.cpp:
```diff
@@ -368,6 +369,51 @@ std::string PPCDebugInterface::GetDescription(u32 address) const
+  std::regex re(",[^r0-]*(-?)(0[xX]?[0-9a-fA-F]*|r\\d+)[^r^s]*.(p|toc|\\d+)");
+  std::smatch match;
+  if (!std::regex_search(instruction, match, re))
+    return 0;
```
Not sure when it would be 0, but optional would be more fitting for an error. I'll change it.

TryTwo replied:
I changed this but someone missed putting it in my recent update.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp:
```diff
@@ -0,0 +1,773 @@
+  if (m_code_trace.size() >= m_record_limit)
+    return;
+
+  CodeTrace tmp_trace;
```
Yeah this should go with it, but the saved instruction should always be the PC or there'll be errors when calculating memory targets..

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp:
```diff
@@ -0,0 +1,773 @@
+    return;
+  }
+
+  for (auto instr = begin_itr; instr != end_itr; instr++)
```
My only issue is when we use the change range option.  I don't think ranged-based for loops have that option.  I could put a "if not yet at start address -> continue" and "if at end address -> break" detection inside a ranged-based loop, if it's more desirable though.

TryTwo replied:
Since I can't use a ranged-based loop in Backtrace(), should I keep the ForwardTrace() loop using iterators, so they behave the same?

Would it be better to pass iterators for the range to ForwardTrace() and Backtrace() or to pass a std::optional<u32> start and end address if the range is changed then calculate the iterators inside those functions?  I need to calculate them in CodeTraceDialog regardless.

sepalani replied:
If you really can't use ranged-based for loop or if it makes the code harder to grasp, feel free to use normal for loops. I'm fine with iterators passed as parameters as well.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp:
```diff
@@ -0,0 +1,773 @@
+    return;
+  }
+
+  for (auto instr = end_itr - 1; instr != begin_itr; instr--)
```
Yeah. I guess I would make a second UpdateIterator for the reverse iterators?

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp:
```diff
@@ -0,0 +1,773 @@
+
+  // Keep stepping until the end_bp or timeout after five(?) seconds
+  using clock = std::chrono::steady_clock;
+  clock::time_point timeout = clock::now() + std::chrono::seconds(10);
```
There's enough room for it I think.

TryTwo replied:
I'm not sure if we need it anymore. After changing to a simpler struct it can record 200,000 instructions in under 3 seconds. The question becomes how many instructions do we realistically need. In 5 seconds it can do 550,000 instructions, which uses about 20MB of storage. Right now it's limited to 200,000 instructions.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.h:
```diff
@@ -0,0 +1,88 @@
+class QListWidgetItem;
+class QSpinBox;
+
+struct CodeTrace
```
That seems like a good way to do it.  CodeTrace, only being used for the direct recording of what's executed, is not meant to be edited after it's creation, but should be cleared before a second recording. I'm not good at connecting various files from scratch though. Would I construct CodeTrace in its own class and pass a reference to CodeTraceDialog?

sepalani replied:
Storing the CodeTrace class as a private member variable in the dialog/GUI class should be fine, IMO.

TryTwo replied:
Is my method a bit weird for a dedicated file?  Would you expect a recording of executed instructions to first output a TraceOutput like struct, which could then be used to make a CodeTrace to break that down that struct for the logic required in Forward/Back tracking a register/memory? Then have another TraceOutput focusing on the final result?

sepalani replied:
I'm fine with the current `TraceOutput` logic. AFAIK, `TraceOutput` is only used in the GUI logic `CodeTraceDialog::DisplayTrace`. My point was to untie it from the GUI, one way of doing it is to make it a return value of `ForwardTrace`, `BackwardTrace`, `Codepath` (these methods should be moved to their own class as suggested for `CodeTrace`) which are the only places where `TraceOutput` is used.

For instance:
```C++
void CodeTraceDialog::DisplayTrace()
{
// ...
std::vector<TraceOutput> trace_out;
if (m_trace_target->text().isEmpty())
    trace_out = m_code_trace.CodePath(/* with parameters from GUI if needed */);
else if (m_backtrace->isChecked())
    trace_out = m_code_trace.Backtrace(/* with parameters from GUI if needed */);
else
    trace_out = m_code_trace.ForwardTrace(/* with parameters from GUI if needed */);
// ...
}
```

If you think you should rework the `TraceOuput` logic to make it simpler, feel free to do so. Though, as I said, I'm fine with the current logic.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.h:
```diff
@@ -0,0 +1,88 @@
+  bool is_load = false;
+};
+
+struct TraceOutput
```
Hmm this is tricky. TraceOutput is created with multiple variables connected to UI options.  Should ForwardTrace and Backtrace, which uses it to track a register/memory from a recorded CodeTrace, also go in a separate file? In the CodeTrace file?

sepalani replied:
If the TraceOuput is only generated by CodeTrace, it can just be the return value of the CodeTrace member functions (```CodeTrace::ForwardTrace```, ```CodeTrace::BackwardTrace```, ```CodeTrace::Codepath```).

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.h:
```diff
@@ -0,0 +1,88 @@
+  void UpdateBreakpoints();
+  void InfoDisp();
+  void SaveInstruction();
+  bool UpdateIterator(std::vector<CodeTrace>::iterator& begin_itr,
```
Simplified enough to remove it.

TryTwo commented on Source/Core/DolphinQt/Debugger/CodeTraceDialog.cpp:
```diff
@@ -0,0 +1,773 @@
+  // i18n: Here, PC is an acronym for program counter, not personal computer.
+  new QListWidgetItem(
+      QStringLiteral(
+          "Used to track a register or memory address and its uses.\n\nInputs:\nRegister or "
```
I updated it somewhat, still probably a work in progress

TryTwo commented:
It needs a bit more work, but the major changes should be done. I also haven't tried to move it outside a dialog window box.

A major change is using TraceOutput as the main vector for the trace. The "CodeTrace" struct was renamed "InstructionAttributes" and only is calculated when needed - not stored in a vector.  TraceOutput can directly be split by CodeTraceDialog into a displayable result for when no target is being tracked. 

I'm guessing DisplayTrace should be chopping into 2-4 smaller functions. 
Should I route CodeTrace functions through DebugInterface?

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
I don't see any boot-to-pause issue in the current version. I booted two different games to pause, recorded 100,000 instructions twice, then hit play and it worked. 

Originally, I wouldn't have recommend running a trace during a game's loading, rather it was meant to be used in the middle of a normal frame. I was worried rapidly stepping through low-level stuff like interrupts, frame-rate timing checks, or whatnot, could crash it. Not an expert on those things.  

Currently, it's crash free for me even on startup stuff.

As for other UI stuff, I'll put it off until I try to switch it to a full widget I guess.

TryTwo commented:
Turned into a widget. Added picture to first post. Fixed a few more issues.

Not sure how to clear it if the game is changed. I think stopping -> restarting a game shouldn't clear anything.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented:
@TryTwo 
I didn't have reviewed the code yet but here are some UI remarks:
 - The "Output options" group box should probably be renamed to "Result filters" or something along those lines.
 - The "Track Target" button should probably be renamed "Refresh", "Update" or "Apply filters" and moved inside and at the end of this "Result filters" group box.
 - The "Reg or Mem" textbox should be moved in this "Result filters" group box with an appropriate text label.
 - The "Result filters" group box should have its own dropdown/textboxes to change range.
 - Since you can't define the starting point I suggest you remove the first dropdown and move the second one into the recording option group box associated with an appropriate label.

As I'm not a UI guy, feel free to suggest idea if you have any but I think these points will make the Trace feature easier to use/understand.

TryTwo commented:
Not ready for a review, but I updated it and made it compatible with the Autostep PR.  I also tried adding the target register's value to the tracing log for each operation, which works well imo.

When Autostep is done, I can record the autostep to a log, then you can see the history of the autostep in trace widget. It also makes tracing a register straight forward, because you just autostep it to trace it.

I don't remember why I did: `StepOpCode(&sync); SingleStep();` in the original recording. Not sure what the difference between the two is.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
I think there the build errors just require a rebase?

The backend is coming along nicely and I fixed some crashes.  The UI still needs a lot of work. I haven't gone back and implemented the suggestions from before yet.

TryTwo commented:
Got a lot of the UI suggestions in.  Updated the pic in the top post.  The numbers on the far right is the target register's value when the trace was being saved.  I just threw it in the table for testing.

I need to figure out how I want the backend for result filters to work. They only apply when you've traced/tracked a value.  It can either be used when you reprocess/update the trace, or they can be saved to the trace output and update immediately on clicking.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
Hopefully this fixes the build errors?
I'm not sure how to make CodeWidget the top widget from inside other widgets.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
So it now just records all register values and when viewing the trace log, you can replace all registers with their actual value. I consider it fairly useful, as you can now see all the values used in every instruction at a glance.

Added color coding to show the target registers/value, overwritten targets, and values that replace registers.  I also found an easy way to let the user choose the colors.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented:
@TryTwo 
Since the PR https://github.com/dolphin-emu/dolphin/pull/10771 was merged, you will need to rebase your PR.

TryTwo commented:
Thanks for all the help reviewing these!  I think I got it rebased correctly now.

I can run a few more passes over the code to clean it up and update comments + the help dialog.   If anyone wants to test it for usability and bugs that'd help.

sepalani commented:
@dolphin-emu-bot rebuild

TryTwo commented:
I think it's ready for a code review, but no rush.  Still haven't updated help contents.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented on Source/Core/DolphinQt/Debugger/TraceWidget.cpp:
```diff
@@ -0,0 +1,952 @@
+{
+  // i18n: Here, PC is an acronym for program counter, not personal computer.
+  auto* info = new QTableWidgetItem(QStringLiteral(
+      "Used to \"track the value in a target register or memory address and its uses.\n\n"
```
Is `"` a typo?

sepalani commented on Source/Core/DolphinQt/Debugger/TraceWidget.cpp:
```diff
@@ -0,0 +1,952 @@
+  // i18n: Here, PC is an acronym for program counter, not personal computer.
+  auto* info = new QTableWidgetItem(QStringLiteral(
+      "Used to \"track the value in a target register or memory address and its uses.\n\n"
+      "Record Trace: Records each executed instruction while stepping from PC to selected "
+      "Breakpoint. Required before tracking a target. If backtracing, set PC to how far back you "
+      "want to trace to and breakpoint the instruction you want to trace backwards.\n\n"
+      "Register: Input examples: r5, f31, use f for ps registers or 80000000 for memory. Only "
+      "takes one value at a time. Leave blank to view complete code path.\n\n"
+      "Starting Address: Used to change range before tracking a value. Record Trace's starting "
+      "address is always the PC. Can change freely after recording trace.\n\n"
+      "Ending breakpoint: Where the trace will stop. If backtracing, should be the line you want "
+      "to backtrace from.\n\n"
+      "Backtrace: A reverse trace that shows where a value came from, the first output line is "
+      "the "
+      "most recent executed.\n\n"
+      "Verbose: Will record all references to what is being tracked, rather than just where it "
+      "is "
+      "moving to or from.\n\n"
+      "Reset on loopback: Will clear the trace if starting address is looped through, ensuring "
+      "only the final loop to the end breakpoint is recorded.\n\n"
+      "Change Range: Change the start and end points of the trace for tracking. Loops may make "
+      "certain ranges buggy.\n\n"
+      "Track target: Follows the register or memory value through the recorded trace. You don't "
+      "have to record a trace multiple times if the first trace recorded the area of code you "
+      "need. You can change any value or option and press track target again. Changing the "
+      "second "
+      "breakpoint will let you backtrace from a new location."));
```
Some names mentioned here don't seem to match the UI interface, I don't see a `verbose` option for instance.

sepalani commented on Source/Core/DolphinQt/Debugger/TraceWidget.cpp:
```diff
@@ -0,0 +1,952 @@
+      "need. You can change any value or option and press track target again. Changing the "
+      "second "
+      "breakpoint will let you backtrace from a new location."));
+  // QFont fixedfont = QFontDatabase::systemFont(QFontDatabase::FixedFont);
+  // fixedfont.setPointSize(11);
+  // info->setFont(fixedfont);
```
These comments might be removed if not needed.

sepalani commented on Source/Core/Core/Debugger/CodeTrace.cpp:
```diff
@@ -50,6 +59,22 @@ u32 GetMemoryTargetSize(std::string_view instr)
+  }
+
+  // debug value
+  return (double)10.0;
```
Shouldn't an `optional` be used rather than a debug value?

sepalani commented on Source/Core/Core/Debugger/CodeTrace.cpp:
```diff
@@ -50,6 +59,22 @@ u32 GetMemoryTargetSize(std::string_view instr)
+  if (!reg.empty())
+  {
+    if (reg.starts_with("r"))
+      return static_cast<double>(ppc_state.gpr[stoi(std::string{reg.begin() + 1, reg.end()})]);
+    else if (reg.starts_with("f"))
+      return ppc_state.ps[stoi(std::string{reg.begin() + 1, reg.end()})].PS0AsDouble();
```
`std::stoi` should be used instead of `stoi`.

sepalani reviewed (changes requested):
I did some quick testing and started a review of the code. During my testing, I tracked a string value in memory based on a memory BP:
![image](https://github.com/dolphin-emu/dolphin/assets/7890055/6e15fd91-27a7-4717-8563-c12a39d22622)

Some remarks during my testing/review of the code:
 - The `Record trace` button (and other UI elements) should be disabled when they can't be used (i.e. the game is already running or no games are running at all).
 - The hardcoded result limit might be an issue, especially if you're stuck in a long/time-consuming function such as SelectThread.
 - The Trace widget should be cleared when the game is stopped.
 - The Trace widget's table needs proper column names (address, mnemonic/value, operand1-3, ...).
 - Tooltips might be added for a better understanding of some checkboxes, labels, buttons, etc.

You might want to add some use cases somewhere because in its current form I don't it's easily accessible and some feature might be misused/misunderstood. Especially since it's not used in conjunction with the play-pause button.

TryTwo commented:
Thanks! I had some other people test it and the UI definitely needs more work, including the help instructions.

TryTwo commented on Source/Core/Core/Debugger/CodeTrace.h:
```diff
@@ -31,26 +29,37 @@ struct InstructionAttributes
+  u32 address;
   std::optional<u32> memory_target = std::nullopt;
   std::string instruction;
+  std::array<double, 4> value{10, 10, 10, 10};
+  std::array<std::string, 4> regs{"", "", "", ""};
```
Hmm the regs being "" like that is redundant. I should drop it.
The double - value, if I understand correctly, is initialized with some random value unless you write something in. I chose 10's as a debug value, but if you have any suggestions for a final value, that'd be great.

Size 4 is because there are at maximum 4 registers used at once.  Would I just do #DEFINE MAX_REGISTERS 4 at the top? Or use something else?

TryTwo replied:
Double is the maximum size  of a value.  Values can be mixed types like double + u32. "lfd" for example loads a double from a register's u32.  I'm not sure if variant works for that...?

sepalani replied:
> The double - value, if I understand correctly, is initialized with some random value unless you write something in. I chose 10's as a debug value, but if you have any suggestions for a final value, that'd be great.

Assuming that value is used to store the register content (if matching the regex), you could use a `std::optional<double>` instead so it has a double value or no value at all (if no match).



> Size 4 is because there are at maximum 4 registers used at once. Would I just do #DEFINE MAX_REGISTERS 4 at the top? Or use something else?

If I understand correctly, it's the maximum number of operands in the instruction (that can be matched by the regex), so you can define a constexpr variable like `constexpr int MAX_OPERANDS` or `MAX_USED_REGISTER`, to make it more obvious. Moreover, there are, at several places, `for` loops based on `<= 3` which is hardcoded and doesn't depend on the size of your container. So it if changes for whatever reason it might introduce issues and make the code less flexible.

> I may need to change the way this works. The value of each double must correspond to the regs with the same index and the first reg has to have its double value filled last (because we have to step before filling it). A bit tricky.

Have you tried `std::map` or a vector/array of `std::pair<std::string,double>` to keep track of the (double) value associated with its register rather than using two distinct arrays?

TryTwo replied:
Alright, I thought hard and came up with a simpler implementation. Put it in its own commit.

TryTwo commented on Source/Core/DolphinQt/Debugger/TraceWidget.cpp:
```diff
@@ -0,0 +1,952 @@
+  QLabel* record_stop_addr_label = new QLabel(tr("Stop recording at:"));
+  m_record_limit_label = new QLabel(tr("Maximum to record"));
+  m_record_limit_input = new QSpinBox();
+  m_record_limit_input->setMinimum(1000);
+  m_record_limit_input->setMaximum(200000);
+  m_record_limit_input->setValue(10000);
+  m_record_limit_input->setSingleStep(10000);
+  m_record_limit_input->setMinimumSize(70, 0);
```
Yeah. I tried to think about the amount of memory taken up with recording and the max amount of time we wanted it to run.  Eventually you usually hit a long loop of dcbi or bdnz (not sure I got those right) or something you don't want.

For results limit, it gets really laggy if you use a huge number. I figured it'd take forever to flip through too many lines of code and the scrollbar would just start racing through it like crazy anyway. You can use change range to get any segment or tick backtrace to quickly get the last section of code, albeit backwards.

TryTwo commented:
Ignore the errant #Define.  Oops.  I rewrote the register stuff and think it's much nicer now.

TryTwo commented:
Disabled buttons when they can't be used.
Changed record limit to be a duration of seconds.
Column names.
Help button with new text.

Note that hitting a wait function (OSThread) while recording usually traps you there. Not sure I can do anything about it, the running trace itself could be blocking what it's waiting on. This seem to be a problem if randomly pausing then recording a trace. Better to trace from a BP.

TryTwo commented:
Some of the buildbots don't have <ranges> so I switched to index based loops for loops that need to be in either direction plus cover any given range.

Need to write the log index to the table so someone can right click and set it.

TryTwo commented:
Refactored and renamed a bit.  Retooled the coloring options, if they look okay I will save/load the user's settings to file.
